tensorboard/data/server/server.rs

```diff
@@ -0,0 +1,216 @@
+        run.tags.push(data::list_scalars_response::TagEntry {
+            tag_name: "accuracy".to_string(),
+            metadata: Some(data::ScalarMetadata {
+                max_step: 5,
+                max_wall_time: FAKE_START_TIME + 5.0,
+                ..Default::default()
+            }),
+            ..Default::default()
+        });
+        res.runs.push(run);
+        Ok(Response::new(res))
+    }
+
+    async fn read_scalars(
+        &self,
+        _request: Request<data::ReadScalarsRequest>,
+    ) -> Result<Response<data::ReadScalarsResponse>, Status> {
+        let mut res: data::ReadScalarsResponse = Default::default();
+        let mut run: data::read_scalars_response::RunEntry = Default::default();
+        run.run_name = "train".to_string();
+        run.tags.push(data::read_scalars_response::TagEntry {
+            tag_name: "accuracy".to_string(),
+            data: Some(data::ScalarData {
+                step: vec![0, 1, 2, 3, 4],
+                wall_time: (0..=4)
+                    .map(|i| FAKE_START_TIME + ((i + 1) as f64))
+                    .collect(),
+                value: vec![0.1, 0.5, 0.8, 0.9, 0.95],
+                ..Default::default()
+            }),
+            ..Default::default()
+        });
+        res.runs.push(run);
+        Ok(Response::new(res))
+    }
+
+    async fn list_tensors(
+        &self,
+        _request: Request<data::ListTensorsRequest>,
+    ) -> Result<Response<data::ListTensorsResponse>, Status> {
+        Err(Status::unimplemented("not yet implemented"))
+    }
+
+    async fn read_tensors(
+        &self,
+        _request: Request<data::ReadTensorsRequest>,
+    ) -> Result<Response<data::ReadTensorsResponse>, Status> {
+        Err(Status::unimplemented("not yet implemented"))
+    }
+
+    async fn list_blob_sequences(
+        &self,
+        _request: Request<data::ListBlobSequencesRequest>,
+    ) -> Result<Response<data::ListBlobSequencesResponse>, Status> {
+        Err(Status::unimplemented("not yet implemented"))
+    }
+
+    async fn read_blob_sequences(
+        &self,
+        _request: Request<data::ReadBlobSequencesRequest>,
+    ) -> Result<Response<data::ReadBlobSequencesResponse>, Status> {
+        Err(Status::unimplemented("not yet implemented"))
+    }
+
+    type ReadBlobStream =
+        Pin<Box<dyn Stream<Item = Result<data::ReadBlobResponse, Status>> + Send + Sync + 'static>>;
+
+    async fn read_blob(
+        &self,
+        _request: Request<data::ReadBlobRequest>,
+    ) -> Result<Response<Self::ReadBlobStream>, Status> {
+        Err(Status::unimplemented("not yet implemented"))
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    use futures_core::Future;
+    use tokio::runtime::Runtime;
+
+    /// Executes the given future in a new, temporary event loop.
+    fn block_on<F: Future>(future: F) -> F::Output {
+        let mut rt = Runtime::new().unwrap();
+        rt.block_on(future)
+    }
```

Comment:
Thought to self: rustc doesn't allow tests to be `async fn`s, but I wonder if it will be allowed in the future.

Reply:
Yep. (I did try it.)

You should be able to write an async test yourself with a bit of
desugaring:

```rust
#[test]
fn test() {
    async fn body() {
        // put whatever you want here
    }
    let mut rt = Runtime::new().unwrap();
    rt.block_on(body)
    block_on(body())
}
```

For `#[test]` to support that natively, it’d need some way to get ahold
of a runtime. (Rust’s futures are runtime-agnostic.) I wouldn’t be
surprised to see `#[tokio::test]` pop up for that, though.

…oh, look: [`#[tokio::test]` exists][tt], and [it does just that][p].
Neat. I guess that is a bit nicer, so I’ll switch over. Thanks!

[tt]: https://docs.rs/tokio/0.3.4/tokio/attr.test.html
[p]: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=c7b03a872e15525707ada636b44126d4